### PR TITLE
CONTRACTS: detect and handle locals in assigns clause checking for loops

### LINKED ISSUE #6943: Validity checks fail for local variables when abstracted loops are executed more than one time
CBMC version: 5.59.0
Operating system: macOS 12.3.1
Exact command line resulting in the issue:
```
goto-cc main.c
goto-instrument --apply-loop-contracts a.out b.out
goto-instrument --pointer-check b.out c.out
cbmc c.out
```
I ran the above command lines on the following c program
```c
static void foo() {
  unsigned i;

  for (i = 0; i < 16; i++)
    __CPROVER_loop_invariant(1 == 1) { int v = 1; }

}


static void bar() {
  unsigned i;

  for (i = 0; i < 16; i++)
    __CPROVER_loop_invariant(1 == 1) { int v = 1; }

}

int main() {
  bar();
  foo();
  foo();
}
```
The result shows that, although the function ```foo``` and ```bar``` are exactly the same, the validity check of ```v``` failed in the function ```foo``` but not in ```bar```. This is the report
```
** Results:
assign.c function bar
[bar.1] line 13 Check loop invariant before entry: SUCCESS
[bar.2] line 13 Check that loop invariant is preserved: SUCCESS
[bar.3] line 13 Check that loop instrumentation was not truncated: SUCCESS
[bar.assigns.2] line 13 Check that i is valid: SUCCESS
[bar.assigns.3] line 13 Check that i is assignable: SUCCESS
[bar.assigns.1] line 14 Check that v is valid: SUCCESS

assign.c function foo
[foo.1] line 4 Check loop invariant before entry: SUCCESS
[foo.2] line 4 Check that loop invariant is preserved: SUCCESS
[foo.3] line 4 Check that loop instrumentation was not truncated: SUCCESS
[foo.assigns.2] line 4 Check that i is valid: SUCCESS
[foo.assigns.3] line 4 Check that i is assignable: SUCCESS
[foo.assigns.1] line 5 Check that v is valid: FAILURE
```
The validity check checks if ```v``` is a dead object, i.e., 
```
  !(POINTER_OBJECT(&v) == POINTER_OBJECT(__CPROVER_dead_object))
```
The check happens before the loop body where ```v``` is declared. When the abstracted loop is executed for the first time, ```v``` is undeclared and not dead, which is the reason why the check success in ```bar```. At the end of the first abstracted loop,  ```v``` will be dead and set to be a dead object
```
        // 49 file assign.c line 5 function foo
        ASSIGN __CPROVER_dead_object := (side_effect statement="nondet" is_nondet_nullable="1" ? cast(address_of(foo::1::1::1::v), empty*) : __CPROVER_dead_object)
        // 50 file assign.c line 5 function foo
        DEAD foo::1::1::1::v
```
Therefore, at the second time we execute the loop, ```v``` is not re-declared yet and is dead, that is why the check failed in ```foo```.

A possible fix is not to check validity of local variables at the beginning of loops. @remi-delmas-3000 Do you have any thoughts

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
Fixes #6943.

When no assigns clause was provided, we were trying to infer variables touched by the loop and generate an assigns clause from that. Some of these variables could be locals, which have no meaning outside the loop, and we were trying to snapshot these variables outside of the loop. We now detect loop locals and remove them from the inferred frame condition the loop.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [N/A] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/contracts/cfg_info.h [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+      }
+    }
+  }
+
+private:
+  const dirtyt dirty_analysis;
+  std::set<irep_idt> locals;
+};
+
+#endif
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/cfg_info.h]: [minor] missing new line at the end

File: src/goto-instrument/contracts/instrument_spec_assigns.h [hunk 2]
```diff
@@ -203,10 +198,12 @@ class instrument_spec_assignst
   instrument_spec_assignst(
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/instrument_spec_assigns.h:199]: Please also add a `\param` comment above for `_cfg_info`.

File: src/goto-instrument/contracts/instrument_spec_assigns.h [hunk 3]
```diff
@@ -499,6 +479,9 @@ class instrument_spec_assignst
   /// Other functions of the model
   const goto_functionst &functions;
 
+  /// Language mode
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/instrument_spec_assigns.h]: Typo?
```suggestion
  /// CFG-based information
```

File: src/goto-instrument/contracts/contracts.cpp [hunk 4]
```diff
@@ -441,16 +452,14 @@ void code_contractst::check_apply_loop_contracts(
   goto_function.body.destructive_insert(
     loop_head, add_pragma_disable_assigns_check(pre_loop_head_instrs));
 
-  optionalt<cfg_infot> cfg_empty_info;
+  // HERE
+  //  loop_cfg_infot cfg_info(goto_function, loop);
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: I am unclear on this. This is probably a stale comment?
If line 456 is still relevant, could you please expand the comment on line 455?

File: src/goto-instrument/contracts/cfg_info.h [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+      return dirty_analysis.get_dirty_ids().find(ident) !=
+             dirty_analysis.get_dirty_ids().end();
+    else
+      return true;
+  }
+
+private:
+  const dirtyt dirty_analysis;
+  const localst locals;
+  std::set<irep_idt> parameters;
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/cfg_info.h]: We have a hash function for ireps available, so we might as well just make these sets unordered:
```suggestion
  std::unordered_set<irep_idt> parameters;
```

File: src/goto-instrument/contracts/cfg_info.h [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+      else
+      {
+        it++;
+      }
+    }
+  }
+
+private:
+  const dirtyt dirty_analysis;
+  std::set<irep_idt> locals;
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/cfg_info.h]: We have a hash function for ireps available, so we might as well just make these sets unordered:
```suggestion
  std::unordered_set<irep_idt> locals;
```

File: src/goto-instrument/contracts/cfg_info.h [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+// For a loop in a goto_function
+class loop_cfg_infot : public cfg_infot
+{
+public:
+  loop_cfg_infot(goto_functiont &_goto_function, const loopt &loop)
+    : dirty_analysis(_goto_function)
+  {
+    goto_programt::targett it = _goto_function.body.instructions.begin();
+    goto_programt::targett end = _goto_function.body.instructions.end();
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/cfg_info.h]: [minor] Could use `auto` here:
```suggestion
    auto it = _goto_function.body.instructions.begin();
    auto end = _goto_function.body.instructions.end();
```

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
Thanks, Remi. Makes sense to me overall.

A few small comments on missing documentation / minor code changes etc.

File: src/goto-instrument/contracts/cfg_info.h [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+#include <analyses/dirty.h>
+#include <analyses/locals.h>
+#include <goto-instrument/havoc_utils.h>
+#include <goto-instrument/loop_utils.h>
+
+#include <set>
+#include <vector>
+
+/// Stores information about a goto function computed from its CFG,
+/// together with a `target` iterator into the instructions of the function.
```
Comment (remi-delmas-3000) [inline src/goto-instrument/contracts/cfg_info.h]: The iterator has been removed, comment cleanup needed

File: src/goto-instrument/contracts/cfg_info.h (hunk 7 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/cfg_info.h]: for some reason auto results in a type error with `loop.contains(it)` in the next lines, I had to manually add the types

File: src/goto-instrument/contracts/contracts.cpp (hunk 4 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: this is a leftover comment

File: src/goto-instrument/contracts/instrument_spec_assigns.h (hunk 2 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/instrument_spec_assigns.h:199]: thanks

File: src/goto-instrument/contracts/instrument_spec_assigns.h (hunk 3 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/instrument_spec_assigns.h]: thanks

File: src/goto-instrument/contracts/cfg_info.h (hunk 7 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/cfg_info.h]: Ah yes, `loop.contains` expects a `const` iterator, but `begin()` returns a non-`const` one.

We could also use the `const` versions of these `begin()` and `end()` iterators if we are not planning on modifying the iterators:
```suggestion
    auto it = _goto_function.body.instructions.cbegin();
    auto end = _goto_function.body.instructions.cend();
```

File: src/goto-instrument/contracts/cfg_info.h [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+  {
+    goto_programt::targett it = _goto_function.body.instructions.begin();
+    goto_programt::targett end = _goto_function.body.instructions.end();
+    while(it != end)
+    {
+      if(it->is_decl() && loop.contains(it))
+        locals.insert(it->decl_symbol().get_identifier());
+
+      it++;
+    }
```
Comment (qinheping) [inline src/goto-instrument/contracts/cfg_info.h]: We can iterate over only instructions in the loop. 
```suggestion
    for(const auto &t : loop)
    {
      if(t->is_decl())
        locals.insert(t->decl_symbol().get_identifier());
    }
```

File: src/goto-instrument/contracts/cfg_info.h [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+    else
+    {
+      // matches ID_address_of, ID_dereference, etc.
+      return false;
+    }
+  }
+};
+
+// For goto_functions
+class function_cfg_infot : public cfg_infot
```
Comment (feliperodri) [inline src/goto-instrument/contracts/cfg_info.h:109]: nitpick: why not add the implementation of these functions on a separate `.c` file?

**Review by feliperodri — APPROVED** [review]
LGTM (considering all existing comments).

File: src/goto-instrument/contracts/cfg_info.h (hunk 9 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/cfg_info.h]: Thanks !

File: src/goto-instrument/contracts/cfg_info.h (hunk 10 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/cfg_info.h:109]: There's only three methods and two concrete classes, it seemed like overkill

**Comment by remi-delmas-3000** [comment]
all changes made

**Review by SaswatPadhi — APPROVED** [review]
LGTM; thanks Remi!

**Pull request merged**